src/gui/widgets/LcdFloatSpinBox.cpp

```diff
@@ -0,0 +1,254 @@
+ * Boston, MA 02110-1301 USA.
+ *
+ */
+
+#include "LcdFloatSpinBox.h"
+
+#include <cmath>
+#include <QApplication>
+#include <QHBoxLayout>
+#include <QInputDialog>
+#include <QLabel>
+#include <QMouseEvent>
+#include <QPainter>
+#include <QPixmap>
+#include <QStyleOptionFrameV2>
+#include <QVBoxLayout>
+
+#include "CaptionMenu.h"
+#include "embed.h"
+#include "GuiApplication.h"
+#include "gui_templates.h"
+#include "MainWindow.h"
+
+
+LcdFloatSpinBox::LcdFloatSpinBox(int numWhole, int numFrac, QWidget* parent, const QString& name) :
+	FloatModelView(new FloatModel(0, 0, 0, 0, nullptr, name, true), this),
+	m_wholeDisplay(numWhole, parent, name, false),
+	m_fractionDisplay(numFrac, parent, name, true),
+	m_mouseMoving(false),
+	m_intStep(false),
+	m_origMousePos(),
+	m_displayOffset(0)
+{
+	layoutSetup();
+}
+
+
+LcdFloatSpinBox::LcdFloatSpinBox(int numWhole, int numFrac, const QString& style, QWidget* parent, const QString& name) :
+	FloatModelView(new FloatModel(0, 0, 0, 0, nullptr, name, true), this),
+	m_wholeDisplay(numWhole, style, parent, name, false),
+	m_fractionDisplay(numFrac, style, parent, name, true),
+	m_mouseMoving(false),
+	m_intStep(false),
+	m_origMousePos(),
+	m_displayOffset(0)
+{
+	layoutSetup(style);
+}
+
+
+void LcdFloatSpinBox::layoutSetup(const QString &style)
+{
+	// Assemble the LCD parts
+	QHBoxLayout *lcdLayout = new QHBoxLayout();
+
+	m_wholeDisplay.setSeamless(false, true);
+	m_fractionDisplay.setSeamless(true, false);
+
+	lcdLayout->addWidget(&m_wholeDisplay);
+
+	QLabel *dotLabel = new QLabel("", this);
+	QPixmap *dotPixmap = new QPixmap(embed::getIconPixmap(QString("lcd_" + style + "_dot").toUtf8().constData()));
+	dotLabel->setPixmap(dotPixmap->copy(0, 0, dotPixmap->size().width(), dotPixmap->size().height() / 2));
+	lcdLayout->addWidget(dotLabel);
+
+	lcdLayout->addWidget(&m_fractionDisplay);
+
+	lcdLayout->setContentsMargins(0, 0, 0, 0);
+	lcdLayout->setSpacing(0);
+
+	// Add space for label
+	QVBoxLayout *outerLayout = new QVBoxLayout();
+	outerLayout->addLayout(lcdLayout);
+	outerLayout->addSpacing(9);
+	outerLayout->setContentsMargins(0, 0, 0, 0);
+	outerLayout->setSizeConstraint(QLayout::SetFixedSize);
+	this->setLayout(outerLayout);
+}
+
+
+void LcdFloatSpinBox::update()
+{
+	const int whole = static_cast<int>(model()->value());
+	const float fraction = model()->value() - whole;
+	const int intFraction = fraction * std::pow(10.f, m_fractionDisplay.numDigits());
+	m_wholeDisplay.setValue(whole + m_displayOffset);
+	m_fractionDisplay.setValue(intFraction + m_displayOffset);
+
+	QWidget::update();
+}
+
+
+void LcdFloatSpinBox::contextMenuEvent(QContextMenuEvent* event)
+{
+	// for the case, the user clicked right while pressing left mouse-
+	// button, the context-menu appears while mouse-cursor is still hidden
+	// and it isn't shown again until user does something which causes
+	// an QApplication::restoreOverrideCursor()-call...
+	mouseReleaseEvent(nullptr);
```

Comment:
I'm not even sure this is needed anymore. I just tested several LMMS versions and since 1.2.2 the cursor is not being hidden anymore when changing spinbox or knob values by dragging. IIRC, this may have something to do with changes required for MacOS, which severely limited access to mouse cursor in some recent version?

If that is the case, I would simply remove this, and also the `QApplication::setOverrideCursor(Qt::BlankCursor)` call in `mousePressEvent()`, since it does not appear to be doing anything anyway (at least not on my machine).

Reply:
> since 1.2.2 the cursor is not being hidden anymore when changing spinbox or knob values by dragging. IIRC, this may have something to do with changes required for MacOS, which severely limited access to mouse cursor in some recent version?

Yes, you may remove cursor stuff from spin boxes.

Reply:
I removed the overrides (appearance and position) from both `LcdFloatSpinBox` and `LcdSpinBox`. Seems to be working the same as before, but would not bet my life on my desktop environment (before the 1.2.2 change, everything I grabbed in LMMS only moved _up_, regardless of the actual mouse movement) so it probably should be tested by someone else as well.